src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
     public void addAll(List<T> list) {
+        int oldElementDataSize = size;
+        ensureCapacityInternal(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            elementData[oldElementDataSize++] = list.get(i);
+        }
+        size = size + list.size();
+    }
 
+    //////// Methods for addition
+
+    private void ensureCapacityInternal(int minCapacity) {
+        if (minCapacity > elementData.length) {
+            long potentialSize = (long) (size * GROW_MULTIPLIER);
+            if (potentialSize < MAX_ARRAY_SIZE) {
+                grow(minCapacity);
+            } else {
+                throw new ArrayListIndexOutOfBoundsException("Addition of "
+                        + minCapacity + " is to large to perform.");
+            }
+        }
     }
 
+    private void grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (MAX_ARRAY_SIZE - minCapacity < elementData.length) {
+            throw new ArrayListIndexOutOfBoundsException("Addition of "
+                    + minCapacity + " is to large to perform.");
+        }
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        newCapacity = Math.max(minCapacity, newCapacity);
+        elementData = Arrays.copyOf(elementData, newCapacity);
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    /////////////// GET - SET //////////////
+
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        rangeCheck(index);
+        elementData[index] = value;
+    }
 
+    private void rangeCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
     }
 
+    /////////////////////// REMOVE //////////////////////////////////////
+
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T oldValue = (T) elementData[index];
+        int numOfElementsMoved = size - index - 1;
+        if (numOfElementsMoved > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, numOfElementsMoved);
+        }
+        elementData[--size] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
+        boolean success = false;
+        int numMoved;
+        for (int i = 0; i < size; i++) {
+            if (element == null || element.equals(elementData[i])) {
+                numMoved = (size - 1) - i;
+                System.arraycopy(elementData, i + 1, elementData, i, numMoved);
+                elementData[--size] = null;
+                success = true;
+                return element;
+            }
```

Comment:
reuse remove(i)

Reply:
<img width="939" alt="image" src="https://github.com/mate-academy/jv-array-list/assets/39467398/f3c60373-fe04-47ba-9238-013e13ed7a2e">
it does not pass tests. Doesn't this cause recursion?

Reply:
I meant with remove (i). Also it doesn't make sense to return removed object in method where object is removed by object-citeria. It would make sense to make such return in remove(index).